### ISSUE TITLE
BashonWindows (Ubuntu from Win10) not finding openssl

### ISSUE THREAD

**Issue by fishouttah20** [drwetter/testssl.sh#337]
I know it's just been released in beta today, but seeing as you tweeted about it...Cloned testssl.sh directly from the git repository, executing testssl.sh from it's directory and getting "Fatal error:
cannot exec or find any openssl binary"

Tried specifying --openssl=/mnt/c/ssl/tools/testssl.sh/bin but same error. 

By default OpenSSL 1.0.1f is part of the default Ubuntu install and that's not being identified either.

**Comment by teward** [comment]
I should point out something critical: BashOnWindows is *not* Full Ubuntu.  You may have to go hunting for the SSL libraries and have them installed in Bash.

Obligatory quote from the Microsoft blogs (https://blogs.windows.com/buildingapps/2016/03/30/run-bash-on-ubuntu-on-windows/):

> First, this is the first time we’re releasing this technology – **it’s marked as beta for a reason**: We know that **there are some rough edges** and that some things will break! Do not expect every Bash script and tool that you run will work perfectly – **there will be gaps**. 

It's entirely possible the SSL libraries don't exist, but I'm more than 100% certain you have to specify an *actual binary* for use in the `--openssl` argument.  You only pass it the directory, not the full path to the OpenSSL executable for use.  (I pass `/usr/local/bin/openssl_1.0.2` for this, but that's just me, and I do it in the script itself with my own changes)

**Comment by fishouttah20** [comment]
Ah, didn't realize I had to specify the binary, when specifying the openssl.Linux.x86_64 binary I get ./testssl.sh: line 5400: 494 Segmentation fault (core dumped) $OPENSSL version -a 2> /dev/null > /dev/null

Apologies when I mentioned the default Ubuntu install, I meant the one specifically installed as part of the BashforWindows process, which does come with 1.0.1f pre-installed. When specifying this binary, I get further:

```
root@localhost:/mnt/c/ssl/tools/testssl.sh# ./testssl.sh --openssl /usr/bin/openssl www.google.com

###########################################################
    testssl.sh       2.7dev from https://testssl.sh/dev/
    (52bd899 2016-04-04 21:52:57 -- 1.478)

      This program is free software. Distribution and
             modification under GPLv2 permitted.
      USAGE w/o ANY WARRANTY. USE IT AT YOUR OWN RISK!

       Please file bugs @ https://testssl.sh/bugs/

###########################################################

 Using "OpenSSL 1.0.1f 6 Jan 2014" [~115 ciphers]
 on localhost:/usr/bin/openssl
 (built: "Feb 29 18:11:15 2016", platform: "debian-amd64")


socket.c:2447: setsockopt(20, SO_TIMESTAMP) failed: Invalid argument
socket.c:1915: internal_send: 158.43.240.4#53: Invalid argument
socket.c:2447: setsockopt(21, SO_TIMESTAMP) failed: Invalid argument
socket.c:1915: internal_send: 158.43.128.72#53: Invalid argument
socket.c:2447: setsockopt(22, SO_TIMESTAMP) failed: Invalid argument
socket.c:1915: internal_send: 194.168.4.100#53: Invalid argument
socket.c:1915: internal_send: 158.43.240.4#53: Invalid argument
socket.c:1915: internal_send: 158.43.128.72#53: Invalid argument
socket.c:1915: internal_send: 194.168.4.100#53: Invalid argument
Fatal error: No IPv4 address for "www.google.com" available
```

**Comment by drwetter** [comment]
I think the bashonwin10 environment will mature over time. It just has been released. Momentarily there's not so much on the testssl side I can do about it.

For this project  it's good to see MS move. If we are lucky there's another major platform testssl will run on.

If I look at the output above and at the code which has been running until here it's good to see that testssl made it that far. (Hint: if you can convince testssl to use nslookup instead you might get even further)

/Dirk


Dirk

**Comment by fishouttah20** [comment]
when running nslookup by itself I'm seeing similar socket invalid argument errors. Reported to BashonWindows guys (though I'd imagine they were aware of something that fundamental)

**Comment by drwetter** [comment]
This is all interesting. But let's wait until BashonWindows is production ready.

I am closing this for project reasons. Feel free to keep on posting though.

**Comment by fishouttah20** [comment]
Got around to retesting this on win10 14366. nslookup and other socket depending things work, and sure enough, it now works when I use --openssl /usr/bin/openssl ...but of course without sslv3 and only 115 ciphers it's not all that useful. When I point it at the openssl.Linux.x86_64 binary included with testssl.sh I'm still getting:

./testssl.sh --openssl /mnt/c/SSL/tools/testssl.sh/bin/openssl.Linux.x86_64 bank.barclays.co.uk
./testssl.sh: line 6664:  7510 Segmentation fault      (core dumped) $OPENSSL version -a 2> /dev/null > /dev/null
Fatal error:
cannot exec or find any openssl binary 

Any ideas? I'll re-open this properly when Anniversary edition is released and feature is in a production version of win10.

**Comment by fishouttah20** [comment]
teward, you need to sign up for the Win10 Insider Preview program (free) and be running a recent Insider Preview version of Win10, then follow this install guide: https://msdn.microsoft.com/en-us/commandline/wsl/install_guide

I'll try compiling my own version, but as you say, I won't hold my breath.

**Comment by fishouttah20** [comment]
TL;DR - It works (but is sloooooow)

So I got as far as pulling the code from github, configured and then got a "make depend" error as no zlib.h file, a quick google and I needed to install libz-dev. (and apt-get auto installed a different package name automatically that I failed to note). 

make depend/make/make report all succeeded

checking and I have 179 ciphers (doing wc of openssl ciphers command)...hmmm, not 191 but then even on the pre-compiled binaries I only ever got 181 but not sure which 3 are missing from mine.

And it works! ...but fair warning it is REALLY slow. Like really REALLY slow.
win10 bash scan: **12 minutes 8 seconds**
win10 host running ubuntu via virtualbox: **5 minutes 14 seconds**

###########################################################
    testssl.sh       2.8rc1 from https://testssl.sh/dev/
    (018468a 2016-07-09 14:24:38 -- 1.522)

      This program is free software. Distribution and
             modification under GPLv2 permitted.
      USAGE w/o ANY WARRANTY. USE IT AT YOUR OWN RISK!

       Please file bugs @ https://testssl.sh/bugs/

###########################################################

 Using "OpenSSL 1.0.2-chacha (1.0.2i-dev)" [~183 ciphers]
 on (snip):/mnt/c/SSL/tools/openssl/apps/openssl
 (built: "Jul 12 18:55:08 2016", platform: "linux-x86_64")

**Comment by drwetter** [comment]
cool!

Execution time in general could be optimized. Against testssl.sh it's 3 minutes through and feels similar to SSLlabs.
 
But also using Msys it's slower than on Linux.


Wrt ciphers: 193 is currently for the current binaries w kerberos, 179 are standard. Plus 4 GOST ciphers if the engine can find the .so file. The latter is probably missing in your case.

**Comment by fishouttah20** [comment]
Took 8 minutes to scan testssl.sh. So yeah, slower, but still cool it works. 

Thanks for the info on the ciphers, I got the 179 count from the openssl binary but when running testssl.sh it says 183, so I guess it's picking up the GOST ciphers too.

**Comment by fishouttah20** [comment]
I just wanted to update on this. I lost my bash on windows environment so had to reinstall everything. I still have to specify the location of openssl but seem to now be able to use the pre-compiled openssl.Linux.x86_64 binary packaged with testssl.sh instead of having to compile my own. Also, a scan only takes 4 minutes instead of 8. Still not blistering fast, but much closer to useable in anger
